File: frontend/src/services/authService.ts
import apiClient, { clearStoredTokens, setStoredTokens } from './api';
import type { ApiResponse, AuthTokens, LoginRequest, RegisterRequest, User } from '../types';

// ─────────────────────────────────────────────────────────────────
// Auth Service
// ─────────────────────────────────────────────────────────────────

export const authService = {
  async login(credentials: LoginRequest): Promise<{ user: User; tokens: AuthTokens }> {
    const res = await apiClient.post<ApiResponse<{ user: User; tokens: AuthTokens }>>(
      '/auth/login',
      credentials,
    );
    const { user, tokens } = res.data.data;
    setStoredTokens({ accessToken: tokens.accessToken, refreshToken: tokens.refreshToken });
    return { user, tokens };
  },

  async logout(): Promise<void> {
    try {
      await apiClient.post('/auth/logout');
    } finally {
      clearStoredTokens();
    }
  },

  async refreshToken(refreshToken: string): Promise<AuthTokens> {
    const res = await apiClient.post<ApiResponse<AuthTokens>>('/auth/refresh', { refreshToken });
    return res.data.data;
  },

  async getProfile(): Promise<User> {
    const res = await apiClient.get<ApiResponse<User>>('/auth/me');
    return res.data.data;
  },

  async register(data: RegisterRequest): Promise<User> {
    const res = await apiClient.post<ApiResponse<User>>('/auth/register', data);
    return res.data.data;
  },
};

File: backend/src/main/java/com/traydstream/service/AuthService.java
package com.traydstream.service;

import com.traydstream.dto.request.LoginRequest;
import com.traydstream.dto.response.AuthResponse;
import com.traydstream.dto.response.UserResponse;
import com.traydstream.entity.User;
import com.traydstream.exception.AppException;
import com.traydstream.repository.UserRepository;
import com.traydstream.security.JwtTokenProvider;
import com.traydstream.security.UserPrincipal;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final AuthenticationManager authManager;
    private final JwtTokenProvider tokenProvider;
    private final UserRepository userRepository;

    @Transactional(readOnly = true)
    public AuthResponse login(LoginRequest request) {
        Authentication auth = authManager.authenticate(
            new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        UserPrincipal principal = (UserPrincipal) auth.getPrincipal();
        String accessToken = tokenProvider.generateAccessToken(principal);
        String refreshToken = tokenProvider.generateRefreshToken(principal);

        User user = userRepository.findByEmail(principal.getEmail())
            .orElseThrow(() -> new AppException("User not found", HttpStatus.NOT_FOUND));

        log.info("User logged in: {}", user.getEmail());

        return AuthResponse.builder()
            .user(mapToUserResponse(user))
            .tokens(AuthResponse.TokenResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .tokenType("Bearer")
                .expiresIn(tokenProvider.getAccessTokenExpirationMs())
                .build())
            .build();
    }

    @Transactional(readOnly = true)
    public AuthResponse refreshToken(String refreshToken) {
        if (!tokenProvider.validateToken(refreshToken)) {
            throw new AppException("Invalid or expired refresh token", HttpStatus.UNAUTHORIZED);
        }

        Long userId = tokenProvider.getUserIdFromToken(refreshToken);
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new AppException("User not found", HttpStatus.NOT_FOUND));

        UserPrincipal principal = UserPrincipal.from(user);
        String newAccessToken = tokenProvider.generateAccessToken(principal);

        return AuthResponse.builder()
            .user(mapToUserResponse(user))
            .tokens(AuthResponse.TokenResponse.builder()
                .accessToken(newAccessToken)
                .refreshToken(refreshToken)
                .tokenType("Bearer")
                .expiresIn(tokenProvider.getAccessTokenExpirationMs())
                .build())
            .build();
    }

    @Transactional(readOnly = true)
    public UserResponse getProfile(Long userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new AppException("User not found", HttpStatus.NOT_FOUND));
        return mapToUserResponse(user);
    }

    private UserResponse mapToUserResponse(User user) {
        return UserResponse.builder()
            .id(user.getId())
            .email(user.getEmail())
            .firstName(user.getFirstName())
            .lastName(user.getLastName())
            .role(user.getRole())
            .isActive(user.getIsActive())
            .createdAt(user.getCreatedAt())
            .updatedAt(user.getUpdatedAt())
            .build();
    }
}
